python/pyspark/pandas/groupby.py

```diff
@@ -581,6 +581,64 @@ def mean(self, numeric_only: Optional[bool] = True) -> FrameLike:
             F.mean, accepted_spark_types=(NumericType,), bool_to_numeric=True
         )
 
+    # TODO: 'q' accepts list like type
+    def quantile(self, q: float = 0.5, accuracy: int = 10000) -> FrameLike:
+        """
+        Return group values at the given quantile.
+
+        .. note:: `quantile` in pandas-on-Spark are using distributed percentile approximation
+            algorithm unlike pandas, the result might different with pandas in accuracy, also
+            `interpolation` parameters are not supported yet.
+
+        Parameters
+        ----------
+        q : float, default 0.5 (50% quantile)
+            Value between 0 and 1 providing the quantile to compute.
+        accuracy : int, optional
+            Default accuracy of approximation. Larger value means better accuracy.
+            The relative error can be deduced by 1.0 / accuracy.
+            This is a panda-on-Spark specific parameter.
+
+            .. versionadded:: 3.4.0
+
+        Returns
+        -------
+        pyspark.pandas.Series or pyspark.pandas.DataFrame
+
+        See Also
+        --------
+        pyspark.pandas.Series.quantile
+        pyspark.pandas.DataFrame.quantile
+        pyspark.sql.functions.percentile_approx
+
+        Examples
+        --------
+        >>> df = ps.DataFrame([
+        ...     ['a', 1], ['a', 2], ['a', 3],
+        ...     ['b', 1], ['b', 3], ['b', 5]
+        ... ], columns=['key', 'val'])
+
+        Groupby one column and return the quantile of the remaining columns in
+        each group.
+
+        >>> df.groupby('key').quantile()
+             val
+        key
+        a    2
+        b    3
```

Comment:
qq: Seems like pandas returns different result as below:
```python
>>> df = pd.DataFrame([
...     ['a', 1], ['a', 2], ['a', 3],
...     ['b', 1], ['b', 3], ['b', 5]
... ], columns=['key', 'val'])
>>> df.groupby('key').quantile()
     val
key
a    2.0
b    3.0
```

Do we want to match the behavior, or is there any reason for this difference ?

Reply:
good catch, I think we should always cast numbers to floats

Reply:
strange, the output of `percentile_approx` should already be floats

Reply:
`return F.percentile_approx(spark_column.cast(DoubleType()), qq, accuracy)` in `DataFrame.quantile`

Reply:
This is a good question! As I mentioned in note: "`quantile` in pandas-on-Spark are using distributed percentile approximation algorithm unlike pandas, the result might different with pandas"

The default pandas quantile is using `linear` interpolation:
```
interpolation{‘linear’, ‘lower’, ‘higher’, ‘midpoint’, ‘nearest’}
This optional parameter specifies the interpolation method to use, when the desired quantile lies between two data points i and j:
linear: i + (j - i) * fraction, where fraction is the fractional part of the index surrounded by i and j.
lower: i.
higher: j.
nearest: i or j whichever is nearest.
midpoint: (i + j) / 2.
```

By default is the `i + (j - i) * fraction`, it is `float` in most cases, so pandas choose to return `float` but for pandas-on-Spark is using distributed percentile approximation, therefore, generally consistent with the input data type.

Reply:
Cool. Then maybe we can keep this behavior as is.

Can you elaborate a bit more in description with what you wrote in this thread ??

Reply:
Of course, if everyone thinks `DoubleType` are correct, I also think it's acceptable

Reply:
Oh, I mean we don't necessarily return the `DoubleType` if there is a reason that our behavior is different from pandas.

I think we can merge as is.

Reply:
Emm, consider df.quantile already force covert the results to double type, let's just keep consist!

@itholic @zhengruifeng Addressed in latest commit! Thanks all!